// gdc-common-utils-ts/dist/models/interoperable-claims/appointment-claims.js
// src/models/fhir/Appointment.claims.ts
/**
 * Defines the flat claims structure for a FHIR Appointment resource, separating
 * version-specific data model fields from version-agnostic API search parameters.
 *
 * This model aligns with FHIR R5 conventions (using `note.text` instead of `comment`)
 * for forward compatibility and consistency with other resources like Communication.
 *
 * @basedon https://github.com/Universal-Health-Chain/uhc-fhir-utils-typescript/blob/main/src/models/templates/Appointment.template.model.ts
 */
/**
 * Contains claims representing fields specific to the FHIR R4/R5 data model for an Appointment.
 * Use this when constructing the `credentialSubject` of a Verifiable Credential or the `resource` object.
 *
 * @basedon https://www.hl7.org/fhir/R4/appointment.html
 */
export var AppointmentClaimsFhirR4Core;
(function (AppointmentClaimsFhirR4Core) {
    AppointmentClaimsFhirR4Core["Identifier"] = "org.hl7.fhir.r4.core.Appointment.identifier";
    AppointmentClaimsFhirR4Core["Status"] = "org.hl7.fhir.r4.core.Appointment.status";
    AppointmentClaimsFhirR4Core["CancelationReason"] = "org.hl7.fhir.r4.core.Appointment.cancelation-reason";
    AppointmentClaimsFhirR4Core["ServiceCategory"] = "org.hl7.fhir.r4.core.Appointment.service-category";
    AppointmentClaimsFhirR4Core["ServiceType"] = "org.hl7.fhir.r4.core.Appointment.service-type";
    AppointmentClaimsFhirR4Core["Specialty"] = "org.hl7.fhir.r4.core.Appointment.specialty";
    AppointmentClaimsFhirR4Core["AppointmentType"] = "org.hl7.fhir.r4.core.Appointment.appointment-type";
    AppointmentClaimsFhirR4Core["ReasonCode"] = "org.hl7.fhir.r4.core.Appointment.reason-code";
    AppointmentClaimsFhirR4Core["ReasonReference"] = "org.hl7.fhir.r4.core.Appointment.reason-reference";
    AppointmentClaimsFhirR4Core["Priority"] = "org.hl7.fhir.r4.core.Appointment.priority";
    AppointmentClaimsFhirR4Core["Description"] = "org.hl7.fhir.r4.core.Appointment.description";
    AppointmentClaimsFhirR4Core["SupportingInformation"] = "org.hl7.fhir.r4.core.Appointment.supporting-information";
    AppointmentClaimsFhirR4Core["Start"] = "org.hl7.fhir.r4.core.Appointment.start";
    AppointmentClaimsFhirR4Core["End"] = "org.hl7.fhir.r4.core.Appointment.end";
    AppointmentClaimsFhirR4Core["MinutesDuration"] = "org.hl7.fhir.r4.core.Appointment.minutes-duration";
    AppointmentClaimsFhirR4Core["Slot"] = "org.hl7.fhir.r4.core.Appointment.slot";
    AppointmentClaimsFhirR4Core["Created"] = "org.hl7.fhir.r4.core.Appointment.created";
    // Comment = 'org.hl7.fhir.r4.core.Appointment.comment', // Deprecated in R5, using note.text for forward compatibility
    AppointmentClaimsFhirR4Core["NoteText"] = "org.hl7.fhir.r4.core.Appointment.note.text";
    AppointmentClaimsFhirR4Core["PatientInstruction"] = "org.hl7.fhir.r4.core.Appointment.patient-instruction";
    AppointmentClaimsFhirR4Core["BasedOn"] = "org.hl7.fhir.r4.core.Appointment.based-on";
    AppointmentClaimsFhirR4Core["ParticipantActor"] = "org.hl7.fhir.r4.core.Appointment.participant.actor";
    AppointmentClaimsFhirR4Core["ParticipantRequired"] = "org.hl7.fhir.r4.core.Appointment.participant.required";
    AppointmentClaimsFhirR4Core["ParticipantStatus"] = "org.hl7.fhir.r4.core.Appointment.participant.status";
    AppointmentClaimsFhirR4Core["ParticipantType"] = "org.hl7.fhir.r4.core.Appointment.participant.type";
    AppointmentClaimsFhirR4Core["RequestedPeriod"] = "org.hl7.fhir.r4.core.Appointment.requested-period";
    AppointmentClaimsFhirR4Core["ResourceId"] = "org.hl7.fhir.r4.core.Appointment.resource.id";
})(AppointmentClaimsFhirR4Core || (AppointmentClaimsFhirR4Core = {}));
/**
 * Contains claims representing the standard, version-agnostic API search parameters for an Appointment.
 *
 * @basedon https://www.hl7.org/fhir/appointment.html#search
 */
export var AppointmentClaimsFhirApi;
(function (AppointmentClaimsFhirApi) {
    AppointmentClaimsFhirApi["Actor"] = "org.hl7.fhir.api.Appointment.actor";
    AppointmentClaimsFhirApi["AppointmentType"] = "org.hl7.fhir.api.Appointment.appointment-type";
    AppointmentClaimsFhirApi["BasedOn"] = "org.hl7.fhir.api.Appointment.based-on";
    AppointmentClaimsFhirApi["Date"] = "org.hl7.fhir.api.Appointment.date";
    AppointmentClaimsFhirApi["Identifier"] = "org.hl7.fhir.api.Appointment.identifier";
    AppointmentClaimsFhirApi["Location"] = "org.hl7.fhir.api.Appointment.location";
    AppointmentClaimsFhirApi["PartStatus"] = "org.hl7.fhir.api.Appointment.part-status";
    AppointmentClaimsFhirApi["Patient"] = "org.hl7.fhir.api.Appointment.patient";
    AppointmentClaimsFhirApi["Practitioner"] = "org.hl7.fhir.api.Appointment.practitioner";
    AppointmentClaimsFhirApi["ReasonCode"] = "org.hl7.fhir.api.Appointment.reason-code";
    AppointmentClaimsFhirApi["ReasonReference"] = "org.hl7.fhir.api.Appointment.reason-reference";
    AppointmentClaimsFhirApi["ServiceCategory"] = "org.hl7.fhir.api.Appointment.service-category";
    AppointmentClaimsFhirApi["ServiceType"] = "org.hl7.fhir.api.Appointment.service-type";
    AppointmentClaimsFhirApi["Slot"] = "org.hl7.fhir.api.Appointment.slot";
    AppointmentClaimsFhirApi["Specialty"] = "org.hl7.fhir.api.Appointment.specialty";
    AppointmentClaimsFhirApi["Status"] = "org.hl7.fhir.api.Appointment.status";
    AppointmentClaimsFhirApi["SupportingInfo"] = "org.hl7.fhir.api.Appointment.supporting-info";
})(AppointmentClaimsFhirApi || (AppointmentClaimsFhirApi = {}));
/**
 * Extends the standard API search parameters with custom parameters derived from core data model fields.
 */
export var AppointmentClaimsFhirApiExtended;
(function (AppointmentClaimsFhirApiExtended) {
    AppointmentClaimsFhirApiExtended["Actor"] = "org.hl7.fhir.api.Appointment.actor";
    AppointmentClaimsFhirApiExtended["AppointmentType"] = "org.hl7.fhir.api.Appointment.appointment-type";
    AppointmentClaimsFhirApiExtended["BasedOn"] = "org.hl7.fhir.api.Appointment.based-on";
    AppointmentClaimsFhirApiExtended["Date"] = "org.hl7.fhir.api.Appointment.date";
    AppointmentClaimsFhirApiExtended["Identifier"] = "org.hl7.fhir.api.Appointment.identifier";
    AppointmentClaimsFhirApiExtended["Location"] = "org.hl7.fhir.api.Appointment.location";
    AppointmentClaimsFhirApiExtended["PartStatus"] = "org.hl7.fhir.api.Appointment.part-status";
    AppointmentClaimsFhirApiExtended["Patient"] = "org.hl7.fhir.api.Appointment.patient";
    AppointmentClaimsFhirApiExtended["Practitioner"] = "org.hl7.fhir.api.Appointment.practitioner";
    AppointmentClaimsFhirApiExtended["ReasonCode"] = "org.hl7.fhir.api.Appointment.reason-code";
    AppointmentClaimsFhirApiExtended["ReasonReference"] = "org.hl7.fhir.api.Appointment.reason-reference";
    AppointmentClaimsFhirApiExtended["ServiceCategory"] = "org.hl7.fhir.api.Appointment.service-category";
    AppointmentClaimsFhirApiExtended["ServiceType"] = "org.hl7.fhir.api.Appointment.service-type";
    AppointmentClaimsFhirApiExtended["Slot"] = "org.hl7.fhir.api.Appointment.slot";
    AppointmentClaimsFhirApiExtended["Specialty"] = "org.hl7.fhir.api.Appointment.specialty";
    AppointmentClaimsFhirApiExtended["Status"] = "org.hl7.fhir.api.Appointment.status";
    AppointmentClaimsFhirApiExtended["SupportingInfo"] = "org.hl7.fhir.api.Appointment.supporting-info";
    // Extended
    AppointmentClaimsFhirApiExtended["CancelationReason"] = "org.hl7.fhir.api.Appointment.cancelation-reason";
    AppointmentClaimsFhirApiExtended["MinutesDuration"] = "org.hl7.fhir.api.Appointment.minutes-duration";
    AppointmentClaimsFhirApiExtended["Priority"] = "org.hl7.fhir.api.Appointment.priority";
    AppointmentClaimsFhirApiExtended["Created"] = "org.hl7.fhir.api.Appointment.created";
    AppointmentClaimsFhirApiExtended["Description"] = "org.hl7.fhir.api.Appointment.description";
    AppointmentClaimsFhirApiExtended["NoteText"] = "org.hl7.fhir.api.Appointment.note-text";
})(AppointmentClaimsFhirApiExtended || (AppointmentClaimsFhirApiExtended = {}));
export const AppointmentClaimsFhirR4CoreMap = {
    [AppointmentClaimsFhirR4Core.Identifier]: String,
    [AppointmentClaimsFhirR4Core.Status]: String,
    [AppointmentClaimsFhirR4Core.CancelationReason]: String,
    [AppointmentClaimsFhirR4Core.ServiceCategory]: String,
    [AppointmentClaimsFhirR4Core.ServiceType]: String,
    [AppointmentClaimsFhirR4Core.Specialty]: String,
    [AppointmentClaimsFhirR4Core.AppointmentType]: String,
    [AppointmentClaimsFhirR4Core.ReasonCode]: String,
    [AppointmentClaimsFhirR4Core.ReasonReference]: String,
    [AppointmentClaimsFhirR4Core.Priority]: Number,
    [AppointmentClaimsFhirR4Core.Description]: String,
    [AppointmentClaimsFhirR4Core.SupportingInformation]: String,
    [AppointmentClaimsFhirR4Core.Start]: String,
    [AppointmentClaimsFhirR4Core.End]: String,
    [AppointmentClaimsFhirR4Core.MinutesDuration]: Number,
    [AppointmentClaimsFhirR4Core.Slot]: String,
    [AppointmentClaimsFhirR4Core.Created]: String,
    [AppointmentClaimsFhirR4Core.NoteText]: String,
    [AppointmentClaimsFhirR4Core.PatientInstruction]: String,
    [AppointmentClaimsFhirR4Core.BasedOn]: String,
    [AppointmentClaimsFhirR4Core.ParticipantActor]: String,
    [AppointmentClaimsFhirR4Core.ParticipantRequired]: String,
    [AppointmentClaimsFhirR4Core.ParticipantStatus]: String,
    [AppointmentClaimsFhirR4Core.ParticipantType]: String,
    [AppointmentClaimsFhirR4Core.RequestedPeriod]: String,
    [AppointmentClaimsFhirR4Core.ResourceId]: String,
};
export const AppointmentClaimsFhirApiMap = {
    [AppointmentClaimsFhirApi.Actor]: String,
    [AppointmentClaimsFhirApi.AppointmentType]: String,
    [AppointmentClaimsFhirApi.BasedOn]: String,
    [AppointmentClaimsFhirApi.Date]: String,
    [AppointmentClaimsFhirApi.Identifier]: String,
    [AppointmentClaimsFhirApi.Location]: String,
    [AppointmentClaimsFhirApi.PartStatus]: String,
    [AppointmentClaimsFhirApi.Patient]: String,
    [AppointmentClaimsFhirApi.Practitioner]: String,
    [AppointmentClaimsFhirApi.ReasonCode]: String,
    [AppointmentClaimsFhirApi.ReasonReference]: String,
    [AppointmentClaimsFhirApi.ServiceCategory]: String,
    [AppointmentClaimsFhirApi.ServiceType]: String,
    [AppointmentClaimsFhirApi.Slot]: String,
    [AppointmentClaimsFhirApi.Specialty]: String,
    [AppointmentClaimsFhirApi.Status]: String,
    [AppointmentClaimsFhirApi.SupportingInfo]: String,
};
export const AppointmentClaimsFhirApiExtendedMap = {
    ...AppointmentClaimsFhirApiMap,
    [AppointmentClaimsFhirApiExtended.CancelationReason]: String,
    [AppointmentClaimsFhirApiExtended.MinutesDuration]: Number,
    [AppointmentClaimsFhirApiExtended.Priority]: Number,
    [AppointmentClaimsFhirApiExtended.Created]: String,
    [AppointmentClaimsFhirApiExtended.Description]: String,
    [AppointmentClaimsFhirApiExtended.NoteText]: String,
};
// src/models/fhir/Appointment.values.ts
/**
 * Defines the possible values for the `status` field on a FHIR Appointment resource.
 * @basedon http://hl7.org/fhir/R4/appointment.html#status
 * @basedon http://hl7.org/fhir/valueset-appointmentstatus.html
 */
export var AppointmentStatus;
(function (AppointmentStatus) {
    AppointmentStatus["Proposed"] = "proposed";
    AppointmentStatus["Pending"] = "pending";
    AppointmentStatus["Booked"] = "booked";
    AppointmentStatus["Arrived"] = "arrived";
    AppointmentStatus["Fulfilled"] = "fulfilled";
    AppointmentStatus["Cancelled"] = "cancelled";
    AppointmentStatus["NoShow"] = "noshow";
    AppointmentStatus["EnteredInError"] = "entered-in-error";
    AppointmentStatus["CheckedIn"] = "checked-in";
    AppointmentStatus["Waitlist"] = "waitlist";
})(AppointmentStatus || (AppointmentStatus = {}));
/**
 * Defines the possible values for the `participant.required` field on a FHIR Appointment resource.
 * @basedon http://hl7.org/fhir/R4/appointment.html#participant
 * @basedon http://hl7.org/fhir/valueset-participantrequired.html
 */
export var AppointmentParticipantRequired;
(function (AppointmentParticipantRequired) {
    AppointmentParticipantRequired["Required"] = "required";
    AppointmentParticipantRequired["Optional"] = "optional";
    AppointmentParticipantRequired["InformationOnly"] = "information-only";
})(AppointmentParticipantRequired || (AppointmentParticipantRequired = {}));
/**
 * Defines the possible values for the `participant.status` field on a FHIR Appointment resource.
 * @basedon http://hl7.org/fhir/R4/appointment.html#participant
 * @basedon http://hl7.org/fhir/valueset-participationstatus.html
 */
export var AppointmentParticipantStatus;
(function (AppointmentParticipantStatus) {
    AppointmentParticipantStatus["Accepted"] = "accepted";
    AppointmentParticipantStatus["Declined"] = "declined";
    AppointmentParticipantStatus["Tentative"] = "tentative";
    AppointmentParticipantStatus["NeedsAction"] = "needs-action";
})(AppointmentParticipantStatus || (AppointmentParticipantStatus = {}));
