// gdc-common-utils-ts/dist/models/interoperable-claims/medication-statement-claims.js
// src/models/fhir/MedicationStatement.claims.ts
/**
 * Flat claims contract for MedicationStatement using FHIR API-like search params.
 *
 * Convention:
 * - Key format: `org.hl7.fhir.api.MedicationStatement.<concrete-parameter>`
 * - `<concrete-parameter>` in kebab-case (no nested dots).
 * - Keep values scalar/string-friendly so they can be persisted as tabular columns.
 */
/**
 * Standard FHIR search parameters for MedicationStatement.
 * @basedon https://hl7.org/fhir/medicationstatement.html#search
 */
export var MedicationStatementClaimsFhirApi;
(function (MedicationStatementClaimsFhirApi) {
    MedicationStatementClaimsFhirApi["Category"] = "org.hl7.fhir.api.MedicationStatement.category";
    MedicationStatementClaimsFhirApi["Code"] = "org.hl7.fhir.api.MedicationStatement.code";
    MedicationStatementClaimsFhirApi["Effective"] = "org.hl7.fhir.api.MedicationStatement.effective";
    MedicationStatementClaimsFhirApi["Identifier"] = "org.hl7.fhir.api.MedicationStatement.identifier";
    MedicationStatementClaimsFhirApi["Medication"] = "org.hl7.fhir.api.MedicationStatement.medication";
    MedicationStatementClaimsFhirApi["PartOf"] = "org.hl7.fhir.api.MedicationStatement.part-of";
    MedicationStatementClaimsFhirApi["Patient"] = "org.hl7.fhir.api.MedicationStatement.patient";
    MedicationStatementClaimsFhirApi["Source"] = "org.hl7.fhir.api.MedicationStatement.source";
    MedicationStatementClaimsFhirApi["Status"] = "org.hl7.fhir.api.MedicationStatement.status";
    MedicationStatementClaimsFhirApi["Subject"] = "org.hl7.fhir.api.MedicationStatement.subject";
})(MedicationStatementClaimsFhirApi || (MedicationStatementClaimsFhirApi = {}));
/**
 * Extended flat parameters used for dosage/timing and operational indexing.
 * These are intentionally scalarized to map cleanly to SQL columns.
 */
export var MedicationStatementClaimsFhirApiExtended;
(function (MedicationStatementClaimsFhirApiExtended) {
    MedicationStatementClaimsFhirApiExtended["Category"] = "org.hl7.fhir.api.MedicationStatement.category";
    MedicationStatementClaimsFhirApiExtended["Code"] = "org.hl7.fhir.api.MedicationStatement.code";
    MedicationStatementClaimsFhirApiExtended["Effective"] = "org.hl7.fhir.api.MedicationStatement.effective";
    MedicationStatementClaimsFhirApiExtended["Identifier"] = "org.hl7.fhir.api.MedicationStatement.identifier";
    MedicationStatementClaimsFhirApiExtended["Medication"] = "org.hl7.fhir.api.MedicationStatement.medication";
    MedicationStatementClaimsFhirApiExtended["PartOf"] = "org.hl7.fhir.api.MedicationStatement.part-of";
    MedicationStatementClaimsFhirApiExtended["Patient"] = "org.hl7.fhir.api.MedicationStatement.patient";
    MedicationStatementClaimsFhirApiExtended["Source"] = "org.hl7.fhir.api.MedicationStatement.source";
    MedicationStatementClaimsFhirApiExtended["Status"] = "org.hl7.fhir.api.MedicationStatement.status";
    MedicationStatementClaimsFhirApiExtended["Subject"] = "org.hl7.fhir.api.MedicationStatement.subject";
    // Dosage-level quantities
    MedicationStatementClaimsFhirApiExtended["DoseQuantity"] = "org.hl7.fhir.api.MedicationStatement.dose-quantity";
    MedicationStatementClaimsFhirApiExtended["DoseQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.dose-quantity-value";
    MedicationStatementClaimsFhirApiExtended["DoseQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.dose-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["RateQuantity"] = "org.hl7.fhir.api.MedicationStatement.rate-quantity";
    MedicationStatementClaimsFhirApiExtended["RateQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.rate-quantity-value";
    MedicationStatementClaimsFhirApiExtended["RateQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.rate-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["RateRangeLowQuantity"] = "org.hl7.fhir.api.MedicationStatement.rate-range-low-quantity";
    MedicationStatementClaimsFhirApiExtended["RateRangeLowQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.rate-range-low-quantity-value";
    MedicationStatementClaimsFhirApiExtended["RateRangeLowQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.rate-range-low-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["RateRangeHighQuantity"] = "org.hl7.fhir.api.MedicationStatement.rate-range-high-quantity";
    MedicationStatementClaimsFhirApiExtended["RateRangeHighQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.rate-range-high-quantity-value";
    MedicationStatementClaimsFhirApiExtended["RateRangeHighQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.rate-range-high-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["RateRatioNumeratorQuantity"] = "org.hl7.fhir.api.MedicationStatement.rate-ratio-numerator-quantity";
    MedicationStatementClaimsFhirApiExtended["RateRatioNumeratorQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.rate-ratio-numerator-quantity-value";
    MedicationStatementClaimsFhirApiExtended["RateRatioNumeratorQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.rate-ratio-numerator-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["RateDenominatorQuantity"] = "org.hl7.fhir.api.MedicationStatement.rate-denominator-quantity";
    MedicationStatementClaimsFhirApiExtended["RateDenominatorQuantityValue"] = "org.hl7.fhir.api.MedicationStatement.rate-denominator-quantity-value";
    MedicationStatementClaimsFhirApiExtended["RateDenominatorQuantityUnit"] = "org.hl7.fhir.api.MedicationStatement.rate-denominator-quantity-unit";
    MedicationStatementClaimsFhirApiExtended["DoseType"] = "org.hl7.fhir.api.MedicationStatement.dose-type";
    MedicationStatementClaimsFhirApiExtended["DosageMethod"] = "org.hl7.fhir.api.MedicationStatement.dosage-method";
    MedicationStatementClaimsFhirApiExtended["DosageMethodText"] = "org.hl7.fhir.api.MedicationStatement.dosage-method-text";
    MedicationStatementClaimsFhirApiExtended["DosageRoute"] = "org.hl7.fhir.api.MedicationStatement.dosage-route";
    MedicationStatementClaimsFhirApiExtended["DosageRouteText"] = "org.hl7.fhir.api.MedicationStatement.dosage-route-text";
    MedicationStatementClaimsFhirApiExtended["DosageSite"] = "org.hl7.fhir.api.MedicationStatement.dosage-site";
    MedicationStatementClaimsFhirApiExtended["DosageSiteText"] = "org.hl7.fhir.api.MedicationStatement.dosage-site-text";
    MedicationStatementClaimsFhirApiExtended["DosageCondition"] = "org.hl7.fhir.api.MedicationStatement.dosage-condition";
    MedicationStatementClaimsFhirApiExtended["DosageAsNeeded"] = "org.hl7.fhir.api.MedicationStatement.dosage-asneeded";
    MedicationStatementClaimsFhirApiExtended["DosagePatientInstructionText"] = "org.hl7.fhir.api.MedicationStatement.dosage-patientinstruction-text";
    MedicationStatementClaimsFhirApiExtended["DoseMaxPerPeriod"] = "org.hl7.fhir.api.MedicationStatement.dose-maxperperiod";
    MedicationStatementClaimsFhirApiExtended["DoseMaxPerAdministration"] = "org.hl7.fhir.api.MedicationStatement.dose-maxperadministration";
    MedicationStatementClaimsFhirApiExtended["DoseMaxPerLifetime"] = "org.hl7.fhir.api.MedicationStatement.dose-maxperlifetime";
    // Timing.repeat scalarization
    MedicationStatementClaimsFhirApiExtended["TimingDescription"] = "org.hl7.fhir.api.MedicationStatement.timing-description";
    MedicationStatementClaimsFhirApiExtended["TimingCount"] = "org.hl7.fhir.api.MedicationStatement.timing-count";
    MedicationStatementClaimsFhirApiExtended["TimingCountMax"] = "org.hl7.fhir.api.MedicationStatement.timing-countmax";
    MedicationStatementClaimsFhirApiExtended["TimingDuration"] = "org.hl7.fhir.api.MedicationStatement.timing-duration";
    MedicationStatementClaimsFhirApiExtended["TimingDurationMax"] = "org.hl7.fhir.api.MedicationStatement.timing-durationmax";
    MedicationStatementClaimsFhirApiExtended["TimingDurationUnit"] = "org.hl7.fhir.api.MedicationStatement.timing-duration-unit";
    MedicationStatementClaimsFhirApiExtended["TimingFrequency"] = "org.hl7.fhir.api.MedicationStatement.timing-frequency";
    MedicationStatementClaimsFhirApiExtended["TimingFrequencyMax"] = "org.hl7.fhir.api.MedicationStatement.timing-frequencymax";
    MedicationStatementClaimsFhirApiExtended["TimingPeriod"] = "org.hl7.fhir.api.MedicationStatement.timing-period";
    MedicationStatementClaimsFhirApiExtended["TimingPeriodMax"] = "org.hl7.fhir.api.MedicationStatement.timing-periodmax";
    MedicationStatementClaimsFhirApiExtended["TimingPeriodUnit"] = "org.hl7.fhir.api.MedicationStatement.timing-period-unit";
    MedicationStatementClaimsFhirApiExtended["TimingDayOfWeek"] = "org.hl7.fhir.api.MedicationStatement.timing-dayofweek";
    MedicationStatementClaimsFhirApiExtended["TimingTimeOfDay"] = "org.hl7.fhir.api.MedicationStatement.timing-timeofday";
    MedicationStatementClaimsFhirApiExtended["TimingWhen"] = "org.hl7.fhir.api.MedicationStatement.timing-when";
    MedicationStatementClaimsFhirApiExtended["TimingOffset"] = "org.hl7.fhir.api.MedicationStatement.timing-offset";
    MedicationStatementClaimsFhirApiExtended["TimingStartOffset"] = "org.hl7.fhir.api.MedicationStatement.timing-startoffset";
    MedicationStatementClaimsFhirApiExtended["TimingEndOffset"] = "org.hl7.fhir.api.MedicationStatement.timing-endoffset";
    // Timing.repeat.bounds[x]
    MedicationStatementClaimsFhirApiExtended["TimingBoundsType"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-type";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsPeriodStart"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-period-start";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsPeriodEnd"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-period-end";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsDuration"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-duration";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsDurationValue"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-duration-value";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsDurationUnit"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-duration-unit";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeLow"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-low";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeLowValue"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-low-value";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeLowUnit"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-low-unit";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeHigh"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-high";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeHighValue"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-high-value";
    MedicationStatementClaimsFhirApiExtended["TimingBoundsRangeHighUnit"] = "org.hl7.fhir.api.MedicationStatement.timing-bounds-range-high-unit";
})(MedicationStatementClaimsFhirApiExtended || (MedicationStatementClaimsFhirApiExtended = {}));
export const MedicationStatementClaimsFhirApiMap = {
    [MedicationStatementClaimsFhirApi.Category]: String,
    [MedicationStatementClaimsFhirApi.Code]: String,
    [MedicationStatementClaimsFhirApi.Effective]: String,
    [MedicationStatementClaimsFhirApi.Identifier]: String,
    [MedicationStatementClaimsFhirApi.Medication]: String,
    [MedicationStatementClaimsFhirApi.PartOf]: String,
    [MedicationStatementClaimsFhirApi.Patient]: String,
    [MedicationStatementClaimsFhirApi.Source]: String,
    [MedicationStatementClaimsFhirApi.Status]: String,
    [MedicationStatementClaimsFhirApi.Subject]: String,
};
export const MedicationStatementClaimsFhirApiExtendedMap = {
    ...MedicationStatementClaimsFhirApiMap,
    [MedicationStatementClaimsFhirApiExtended.DoseQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.DoseQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.DoseQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.RateQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.RateQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.RateQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRangeLowQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRangeLowQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.RateRangeLowQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRangeHighQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRangeHighQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.RateRangeHighQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRatioNumeratorQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.RateRatioNumeratorQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.RateRatioNumeratorQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.RateDenominatorQuantity]: String,
    [MedicationStatementClaimsFhirApiExtended.RateDenominatorQuantityValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.RateDenominatorQuantityUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.DoseType]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageMethod]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageMethodText]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageRoute]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageRouteText]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageSite]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageSiteText]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageCondition]: String,
    [MedicationStatementClaimsFhirApiExtended.DosageAsNeeded]: Boolean,
    [MedicationStatementClaimsFhirApiExtended.DosagePatientInstructionText]: String,
    [MedicationStatementClaimsFhirApiExtended.DoseMaxPerPeriod]: String,
    [MedicationStatementClaimsFhirApiExtended.DoseMaxPerAdministration]: String,
    [MedicationStatementClaimsFhirApiExtended.DoseMaxPerLifetime]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingDescription]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingCount]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingCountMax]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingDuration]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingDurationMax]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingDurationUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingFrequency]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingFrequencyMax]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingPeriod]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingPeriodMax]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingPeriodUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingDayOfWeek]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingTimeOfDay]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingWhen]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingOffset]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingStartOffset]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingEndOffset]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsType]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsPeriodStart]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsPeriodEnd]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsDuration]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsDurationValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsDurationUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeLow]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeLowValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeLowUnit]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeHigh]: String,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeHighValue]: Number,
    [MedicationStatementClaimsFhirApiExtended.TimingBoundsRangeHighUnit]: String,
};
