// gdc-common-utils-ts/dist/constants/Schemas.js
// constants/Schemas.ts
// Based on the backend's src/models/schemaorg.ts
/**
 * Defines the types of form requests that can be sent in a batch.
 */
export const FormRequestType = {
    IndividualTerms: 'IndividualTerms',
    PersonalIdentity: 'PersonalIdentity',
};
// --- Enums for URL construction and validation ---
export const Sector = {
    EMERGENCY: 'emergency',
    HEALTH_CARE: 'health-care',
    HEALTH_INSURANCE: 'health-insurance',
    RESEARCH: 'research',
};
export const Section = {
    REGISTRY: 'registry',
    ENTITY: 'entity',
    INDIVIDUAL: 'individual',
    NETWORK: 'network',
};
export const Format = {
    SCHEMA: 'org.schema',
    FHIR_API: 'org.hl7.fhir.api',
};
export const Resource = {
    PERSON: 'Person',
    RELATED_PERSON: 'RelatedPerson',
    EMPLOYEE: 'Employee',
    EMPLOYEE_ROLE: 'EmployeeRole',
    PRACTITIONER: 'Practitioner',
    PRACTITIONER_ROLE: 'PractitionerRole',
    ORGANIZATION: 'Organization',
    LOCATION: 'Location',
    GROUP: 'Group',
};
export const JobAction = {
    BATCH: '_batch',
    CREATE: '_create',
    DISCOVERY: '_discovery',
};
export const knownDomainsReversed = [
    'org.schema',
    'org.hl7.fhir',
    'org.ilo.isco',
    'net.openid',
];
