// gdc-common-utils-ts/dist/constants/schemaorg.js
// Copyright 2025 Antifraud Services Inc. under the Apache License, Version 2.0.
// File: src/models/schemaorg.ts
export var ClaimsServiceSchemaorg;
(function (ClaimsServiceSchemaorg) {
    ClaimsServiceSchemaorg["category"] = "org.schema.Service.category";
    ClaimsServiceSchemaorg["identifier"] = "org.schema.Service.identifier";
    ClaimsServiceSchemaorg["serviceType"] = "org.schema.Service.serviceType";
    ClaimsServiceSchemaorg["termsOfService"] = "org.schema.Service.termsOfService";
})(ClaimsServiceSchemaorg || (ClaimsServiceSchemaorg = {}));
/**
 * Defines the canonical claim names for the 'org.schema' context,
 * based on Schema.org vocabulary.
 */
export var ClaimsOrganizationSchemaorg;
(function (ClaimsOrganizationSchemaorg) {
    /** ISO 3166-1 alpha-2 (two-letter country code). The jurisdiction could be the country or the region (county, province or state) */
    ClaimsOrganizationSchemaorg["addressCountry"] = "org.schema.Organization.address.addressCountry";
    /** ISO 3166-2 code for administrative divisions. The jurisdiction could be the country or the region (county, province or state) */
    ClaimsOrganizationSchemaorg["addressRegion"] = "org.schema.Organization.address.addressRegion";
    ClaimsOrganizationSchemaorg["addressLocality"] = "org.schema.Organization.address.addressLocality";
    ClaimsOrganizationSchemaorg["postalCode"] = "org.schema.Organization.address.postalCode";
    ClaimsOrganizationSchemaorg["streetAddress"] = "org.schema.Organization.address.streetAddress";
    /** `TAX` ID or `EI` (Employer ID): @see http://terminology.hl7.org/CodeSystem/v2-0203 */
    ClaimsOrganizationSchemaorg["identifierType"] = "org.schema.Organization.identifier.additionalType";
    ClaimsOrganizationSchemaorg["identifierValue"] = "org.schema.Organization.identifier.value";
    /** Legal registered name */
    ClaimsOrganizationSchemaorg["legalName"] = "org.schema.Organization.legalName";
    /** Commercial name */
    ClaimsOrganizationSchemaorg["name"] = "org.schema.Organization.name";
    /** short url-friendly name (0-9,a-z) */
    ClaimsOrganizationSchemaorg["alternateName"] = "org.schema.Organization.alternateName";
    /** External URL for the service */
    ClaimsOrganizationSchemaorg["url"] = "org.schema.Organization.url";
    /** The identifier is a URN generated using the legal ID (TAX or EI) */
    ClaimsOrganizationSchemaorg["identifier"] = "org.schema.Organization.identifier";
    /** DUNS (free) or LEI could be provided */
    ClaimsOrganizationSchemaorg["duns"] = "org.schema.Organization.duns";
    /** Public contact email */
    ClaimsOrganizationSchemaorg["email"] = "org.schema.Organization.email";
    /** Public contact phone */
    ClaimsOrganizationSchemaorg["telephone"] = "org.schema.Organization.telephone";
    ClaimsOrganizationSchemaorg["numberOfEmployees"] = "org.schema.Organization.numberOfEmployees.value"; // to purchase licenses for device profile's activation
})(ClaimsOrganizationSchemaorg || (ClaimsOrganizationSchemaorg = {}));
export var ClaimsOfferSchemaorg;
(function (ClaimsOfferSchemaorg) {
    ClaimsOfferSchemaorg["acceptedPaymentMethod"] = "org.schema.Offer.acceptedPaymentMethod";
    ClaimsOfferSchemaorg["category"] = "org.schema.Offer.category";
    ClaimsOfferSchemaorg["checkoutPageURLTemplate"] = "org.schema.Offer.checkoutPageURLTemplate";
    ClaimsOfferSchemaorg["eligibleCustomerType"] = "org.schema.Offer.eligibleCustomerType";
    ClaimsOfferSchemaorg["eligibleQuantityValue"] = "org.schema.Offer.eligibleQuantity.value";
    ClaimsOfferSchemaorg["identifier"] = "org.schema.Offer.identifier";
    ClaimsOfferSchemaorg["itemOfferedName"] = "org.schema.Offer.itemOffered.name";
    ClaimsOfferSchemaorg["itemOfferedSku"] = "org.schema.Offer.itemOffered.sku";
    ClaimsOfferSchemaorg["offeredBy"] = "org.schema.Offer.offeredBy";
    ClaimsOfferSchemaorg["price"] = "org.schema.Offer.price";
    ClaimsOfferSchemaorg["priceCurrency"] = "org.schema.Offer.priceCurrency";
    ClaimsOfferSchemaorg["serialNumber"] = "org.schema.Offer.serialNumber";
})(ClaimsOfferSchemaorg || (ClaimsOfferSchemaorg = {}));
/** For Employees (and Employee Role, but no PII) and customers / related persons.
 * - `givenName`: The given name of the person.
 * - `familyName`: The primary family name or surname of the person.
 * - `alternateName`: The second surname or mother's family name, used for facilitating searches
 *   and catering to cultures with multiple surnames.
 * - `name`: The transliterated full name of the person, useful for standardized naming
 *   conventions and international contexts.
 */
export var ClaimsPersonSchemaorg;
(function (ClaimsPersonSchemaorg) {
    /** Second surname or mother's maiden name */
    ClaimsPersonSchemaorg["additionalName"] = "org.schema.Person.additionalName";
    /** Short friendly name */
    ClaimsPersonSchemaorg["alternateName"] = "org.schema.Person.alternateName";
    ClaimsPersonSchemaorg["birthDate"] = "org.schema.Person.birthDate";
    ClaimsPersonSchemaorg["email"] = "org.schema.Person.email";
    ClaimsPersonSchemaorg["familyName"] = "org.schema.Person.familyName";
    ClaimsPersonSchemaorg["gender"] = "org.schema.Person.gender";
    ClaimsPersonSchemaorg["givenName"] = "org.schema.Person.givenName";
    ClaimsPersonSchemaorg["hasOccupation"] = "org.schema.Person.hasOccupation";
    /** Credential material bound to the person identity in VC payloads (for activation flow this is the representative signing kid). */
    ClaimsPersonSchemaorg["hasCredentialMaterial"] = "org.schema.Person.hasCredential.material";
    ClaimsPersonSchemaorg["identifier"] = "org.schema.Person.identifier";
    ClaimsPersonSchemaorg["identifierType"] = "org.schema.Person.identifier.additionalType";
    ClaimsPersonSchemaorg["identifierValue"] = "org.schema.Person.identifier.value";
    /** ICAO transliteration of official given name (including middlenames), family name and addtional surname */
    ClaimsPersonSchemaorg["name"] = "org.schema.Person.name";
    ClaimsPersonSchemaorg["memberOf"] = "org.schema.Person.memberOf";
    ClaimsPersonSchemaorg["memberOfTaxId"] = "org.schema.Person.memberOf.taxID";
    ClaimsPersonSchemaorg["telephone"] = "org.schema.Person.telephone";
    ClaimsPersonSchemaorg["worksFor"] = "org.schema.Person.worksFor";
    /*
    gender = 'org.schema.Person.gender',
    birthDate = 'org.schema.Person.birthdate',      // Date: Date of birth.
    birthPlace = 'org.schema.Person.birthplace',    // Place: The place where the person was born.
    nationality = 'org.schema.Person.nationality',  // Country: Nationality of the person.
    height = 'org.schema.Person.height',
    // Properties from Thing
    additionalType = 'org.schema.Person.additionaltype', // e.g.: 'Employee'
    */
})(ClaimsPersonSchemaorg || (ClaimsPersonSchemaorg = {}));
/**
 * Defines the flat claim structure for a schema.org/Action.
 * This is used for requests where an entity (agent) performs an action,
 * often with a human controller (participant) initiating it.
 */
export var ClaimsActionSchemaorg;
(function (ClaimsActionSchemaorg) {
    // The primary agent performing the action (e.g., the Tenant Organization)
    ClaimsActionSchemaorg["agentIdentifier"] = "org.schema.Action.agent.identifier";
    ClaimsActionSchemaorg["agentLegalName"] = "org.schema.Action.agent.legalName";
    // ... other flattened properties of the agent ...
    // A co-agent participating in the action (e.g., the Human Controller T)
    ClaimsActionSchemaorg["participantIdentifier"] = "org.schema.Action.participant.identifier";
    // The service provider or target of the action (e.g., the Fabric Network)
    ClaimsActionSchemaorg["providerIdentifier"] = "org.schema.Action.provider.identifier";
    ClaimsActionSchemaorg["providerName"] = "org.schema.Action.provider.name";
    // The time the action was initiated
    ClaimsActionSchemaorg["startTime"] = "org.schema.Action.startTime";
})(ClaimsActionSchemaorg || (ClaimsActionSchemaorg = {}));
export const ICAOReverseDns = 'int.icao';
export var ICAOIdentityParams;
(function (ICAOIdentityParams) {
    ICAOIdentityParams["HairColor"] = "int.icao.mrtd.hair-color";
})(ICAOIdentityParams || (ICAOIdentityParams = {}));
export const indexedPersonAttributeList = [
    ClaimsPersonSchemaorg.givenName,
    ClaimsPersonSchemaorg.familyName,
    ClaimsPersonSchemaorg.additionalName,
    ClaimsPersonSchemaorg.email,
    ClaimsPersonSchemaorg.telephone,
    ClaimsPersonSchemaorg.birthDate,
    ClaimsPersonSchemaorg.identifierType,
    ClaimsPersonSchemaorg.identifierValue,
    // ClaimsPersonSchemaorg.additionalType,
];
export const fullPersonParamsSchemaorg = [
    {
        name: ClaimsPersonSchemaorg.additionalName,
        type: 'string',
        value: undefined,
        unique: true,
    },
    {
        name: ClaimsPersonSchemaorg.familyName,
        type: 'string',
        value: undefined,
        unique: true,
    },
    { name: ClaimsPersonSchemaorg.email, type: 'string', value: undefined },
    {
        name: ClaimsPersonSchemaorg.givenName,
        type: 'string',
        value: undefined,
    },
    {
        name: ClaimsPersonSchemaorg.telephone,
        type: 'string',
        value: undefined,
    },
    {
        name: ClaimsPersonSchemaorg.gender,
        type: 'string',
        value: undefined,
        unique: true,
    },
    {
        name: ClaimsPersonSchemaorg.identifierType,
        type: 'string',
        value: undefined,
    },
    {
        name: ClaimsPersonSchemaorg.identifierValue,
        type: 'string',
        value: undefined,
    },
    {
        name: ClaimsPersonSchemaorg.birthDate,
        type: 'string',
        value: undefined,
        unique: true,
    },
    /*
    {
      name: ClaimsPersonSchemaorg.additionalType,
      type: 'string',
      value: undefined,
    },
    */
];
