// gdc-common-utils-ts/dist/utils/did.js
// crypto-ts/utils/did.ts
// Copyright 2025 Antifraud Services Inc. under the Apache License, Version 2.0.
/**
 * Generates a DID Service ID fragment from a selector object.
 * The format is always `#<section>:<format>:<resourceType>:<action>`.
 *
 * This utility's only job is to correctly assemble the string. It contains no
 * conditional logic about DID types, as that is handled by the data layer.
 *
 * @param selector The structured object containing the endpoint parts.
 * @returns The formatted service ID fragment string.
 */
export function generateServiceId(selector) {
    // Canonical casing:
    // - `section`, `format`, `resourceType` are matched case-insensitively by the backend router,
    //   and lowercasing them makes DID Document lookup deterministic across SDKs.
    // - `action` is kept as-is (future actions may be case-sensitive identifiers).
    const parts = [
        selector.section?.toLowerCase(),
        selector.format?.toLowerCase(),
        selector.resourceType?.toLowerCase(),
        selector.action,
    ];
    return `#${parts.filter(p => p).join(':')}`;
}
/**
 * Normalizes a did:web string to a canonical format to prevent resolution errors
 * due to case sensitivity in the path component of the underlying URL.
 *
 * The rule is:
 * - All segments are lowercased, EXCEPT the final segment.
 * - If the final segment represents a `system|code` pair (e.g., for a role),
 *   the `system` part is lowercased, but the `code` is preserved as-is.
 * - If the final segment is a unique identifier (like a Tax ID), it is preserved as-is.
 *
 * @param did The did:web string to normalize.
 * @returns The canonicalized did:web string.
 */
export function normalizeDidWeb(did) {
    const parts = did.split(':');
    if (parts[0] !== 'did' || parts[1] !== 'web' || parts.length < 3) {
        // Not a valid did:web, return as is.
        return did;
    }
    // Lowercase all parts except the very last one.
    const lowercasedParts = parts.slice(0, -1).map(part => part.toLowerCase());
    let lastPart = parts[parts.length - 1];
    // Special handling for the last part if it contains a role descriptor.
    if (lastPart.includes('|')) {
        const [system, ...codeParts] = lastPart.split('|');
        const code = codeParts.join('|'); // Re-join in case the code itself has a pipe.
        lastPart = `${system.toLowerCase()}|${code}`;
    }
    return [...lowercasedParts, lastPart].join(':');
}
/**
* Encodes a hostname according to did:web spec (percent-encodes port colons).
* @param apiUrl The full base URL of the API.
* @returns The percent-encoded hostname.
*/
function getEncodedHost(apiUrl) {
    try {
        const parsedUrl = new URL(apiUrl);
        return parsedUrl.host.replace(':', '%3A');
    }
    catch (e) {
        console.error(`[getEncodedHost] Invalid apiUrl provided: ${apiUrl}`);
        return 'localhost'; // Fallback
    }
}
/**
* Creates the deterministic "hosted" did:web for a tenant (Organization).
*
* @param hostDidWeb The DID of the host (e.g., 'did:web:host.example.com').
* @param tenantAlternateName The alternate name of the tenant (e.g., 'acme').
* @param context An object containing jurisdiction, version, and sector.
* @returns The tenant's full, correctly formatted hosted did:web.
*          Example: 'did:web:host.example.com:acme:cds-es:v1:health-care'
*/
export function createHostedDidWeb(hostDidWeb, tenantAlternateName, context) {
    const hostPart = hostDidWeb.replace(/^did:web:/, '');
    // The path in a did:web uses colons as separators.
    const didPath = `cds-${context.jurisdiction}:${context.version}:${context.sector}`;
    return `did:web:${hostPart}:${tenantAlternateName}:${didPath}`;
}
/**
 * Builds the canonical details (URL and did:web URN) for a hosted DID.
 * This is the single source of truth for constructing hosted identifiers in the app.
 *
 * @param options An object with the components of the DID.
 * @param options.host The provider's domain (e.g., 'provider.com').
 * @param options.alternateName The tenant's identifier (e.g., 'acme').
 * @param options.jurisdiction The legal jurisdiction (defaults to 'ES').
 * @param options.sector The business sector (defaults to 'health-care').
 * @returns An object with the full URL (with trailing /) and the canonical did:web.
 */
export function buildHostedDidDetails({ host, alternateName, jurisdiction = 'ES', sector = 'health-care', }) {
    // 1. Build the path part of the DID/URL.
    const pathPart = `${alternateName}/cds-${jurisdiction}/v1/${sector}`;
    // 2. Build the full URL, ensuring a trailing slash.
    const url = `https://${host}/${pathPart}/`;
    // 3. Build the canonical did:web, replacing / with :.
    const hostAndPath = `${host}/${pathPart}`;
    const did = `did:web:${hostAndPath.replace(/\//g, ':')}`;
    return { did, url };
}
/**
* Converts a did:web identifier into a full HTTPS or HTTP base URL with a trailing slash.
* It correctly decodes percent-encoded ports for local development.
* @param did The did:web string (e.g., 'did:web:example.com' or 'did:web:localhost%3A3000:acme:cds-es:v1:health-care').
* @returns The full base URL (e.g., 'https://example.com/acme/cds-es/v1/health-care/').
*/
export function getBaseUrlFromDidWeb(did) {
    const didParts = did.replace(/^did:web:/, '').split(':');
    const domainPart = didParts[0];
    const pathParts = didParts.slice(1);
    const decodedDomain = decodeURIComponent(domainPart);
    const protocol = decodedDomain.startsWith('localhost') ? 'http' : 'https';
    const path = pathParts.join('/').replace(/cds-(es|us|gb)/, (match, p1) => `cds-${p1.toUpperCase()}`);
    // Ensure a trailing slash for the base URL, without double slashes when no path is present.
    const normalizedPath = path ? `${path}/` : '';
    return `${protocol}://${decodedDomain}/${normalizedPath}`;
}
