// gdc-common-utils-ts/dist/utils/didcomm-submit.js
import { resolveDidcommSubmissionPlan } from './didcomm-submit-policy.js';
function getHeaderValue(headers, name) {
    if (!headers || typeof headers.get !== 'function') {
        return undefined;
    }
    const value = headers.get(name);
    return value == null ? undefined : value;
}
async function parseBody(response) {
    if (typeof response.text === 'function') {
        const raw = await response.text();
        if (!raw)
            return undefined;
        try {
            return JSON.parse(raw);
        }
        catch {
            return raw;
        }
    }
    return undefined;
}
export async function submitDidcomm(input) {
    const plan = resolveDidcommSubmissionPlan(input.mode, {
        hasRecipientEncryptionJwk: !!input.recipientEncryptionJwk,
    });
    const headers = {
        ...(input.defaultHeaders ?? {}),
        Accept: 'application/json, application/didcomm-plaintext+json, */*',
    };
    let body;
    let contentType;
    if (plan.submitKind === 'plain') {
        contentType = 'application/didcomm-plaintext+json';
        body = JSON.stringify(input.payload);
    }
    else {
        if (!input.signCompactJws) {
            throw new Error('Encrypted DIDComm submission requires signCompactJws callback.');
        }
        if (!input.encryptCompactJwe) {
            throw new Error('Encrypted DIDComm submission requires encryptCompactJwe callback.');
        }
        if (!input.recipientEncryptionJwk) {
            throw new Error('Encrypted DIDComm submission requires recipientEncryptionJwk.');
        }
        const compactJws = await input.signCompactJws(input.payload);
        body = await input.encryptCompactJwe(compactJws, input.recipientEncryptionJwk);
        contentType = 'application/didcomm-encrypted+json';
    }
    headers['Content-Type'] = contentType;
    if (input.bearerToken) {
        headers.Authorization = `Bearer ${input.bearerToken}`;
    }
    const response = await input.fetcher(input.url, {
        method: 'POST',
        headers,
        body,
    });
    return {
        status: response.status,
        location: getHeaderValue(response.headers, 'location'),
        body: await parseBody(response),
        submitKind: plan.submitKind,
        contentType,
    };
}
