// gdc-common-utils-ts/dist/utils/didcomm.js
/**
 * Generates a random ID for DIDComm messages.
 */
function generateId() {
    return Math.random().toString(36).substr(2, 9);
}
/**
 * Creates a DIDComm message instance.
 */
export class DidCommMessage {
    id;
    type;
    from;
    to;
    thid;
    pthid;
    created_time;
    expires_time;
    body;
    attachments;
    constructor() {
        this.id = generateId();
        this.type = '';
        this.body = {};
    }
}
/**
 * Prepares a DIDComm request message.
 */
export function prepareDidCommRequest(type, body = {}, attachments = []) {
    const message = new DidCommMessage();
    message.type = type;
    message.body = body;
    message.attachments = attachments;
    message.thid = message.id; // Set thid to id for new threads
    return message;
}
/**
 * Includes VP token in DIDComm message body.
 */
export function includeVpTokenInMessage(message, vpToken) {
    message.body.vp_token = vpToken;
}
/**
 * Includes file bytes as base64 attachment in DIDComm message.
 */
export function includeFileInMessage(message, fileBytes, mediaType, id) {
    const base64 = Buffer.from(fileBytes).toString('base64');
    if (!message.attachments)
        message.attachments = [];
    message.attachments.push({
        id,
        media_type: mediaType,
        data: { base64 }
    });
}
/**
 * Gets THID from DIDComm message.
 */
export function getThidFromMessage(message) {
    return message.thid || message.id;
}
/**
 * Gets data results from polling response body.
 */
export function getDataResults(response) {
    return response.body?.data || [];
}
