// gdc-common-utils-ts/dist/utils/fhir-cid.js
// Copyright 2026 Antifraud Services Inc. under the Apache License, Version 2.0.
import { sha256 } from '@noble/hashes/sha2.js';
import { utf8ToBytes } from '@noble/hashes/utils.js';
import { encodeMultibase58btc } from './multibase58.js';
const DEFAULT_MULTICODEC_DAG_JSON = 0x0129;
const MULTIHASH_SHA2_256_CODE = 0x12;
const MULTIHASH_SHA2_256_LEN = 32;
const CID_V1 = 0x01;
function toHex(bytes) {
    return Buffer.from(bytes).toString('hex');
}
function encodeVarint(value) {
    if (!Number.isInteger(value) || value < 0) {
        throw new Error(`Invalid varint value: ${value}`);
    }
    const out = [];
    let n = value >>> 0;
    while (n >= 0x80) {
        out.push((n & 0x7f) | 0x80);
        n >>>= 7;
    }
    out.push(n);
    return Uint8Array.from(out);
}
function concatBytes(...parts) {
    const total = parts.reduce((acc, part) => acc + part.length, 0);
    const out = new Uint8Array(total);
    let offset = 0;
    for (const part of parts) {
        out.set(part, offset);
        offset += part.length;
    }
    return out;
}
function canonicalizeValue(value, options, depth) {
    if (Array.isArray(value)) {
        // FHIR array order can be significant; preserve order, canonicalize each element.
        return value.map((entry) => canonicalizeValue(entry, options, depth + 1));
    }
    if (value && typeof value === 'object') {
        const asRecord = value;
        const keys = Object.keys(asRecord).sort();
        const out = {};
        for (const key of keys) {
            if (options.stripNarrativeText && depth === 0 && key === 'text') {
                continue;
            }
            if (options.stripNestedElementIds && depth > 0 && key === 'id') {
                continue;
            }
            let next = asRecord[key];
            if (next === undefined)
                continue;
            if (options.stripMetaVersionId && key === 'meta' && next && typeof next === 'object' && !Array.isArray(next)) {
                const metaRecord = { ...next };
                delete metaRecord.versionId;
                next = metaRecord;
            }
            out[key] = canonicalizeValue(next, options, depth + 1);
        }
        return out;
    }
    return value;
}
export function canonicalizeFhirResource(resource, options = {}) {
    const normalizedOptions = {
        stripMetaVersionId: options.stripMetaVersionId ?? true,
        stripNarrativeText: options.stripNarrativeText ?? false,
        stripNestedElementIds: options.stripNestedElementIds ?? false,
    };
    const normalized = canonicalizeValue(resource, normalizedOptions, 0);
    return JSON.stringify(normalized);
}
export function buildCidV1FromCanonicalJson(canonicalJson, multicodecCode = DEFAULT_MULTICODEC_DAG_JSON) {
    const digest = sha256(utf8ToBytes(canonicalJson));
    const multihash = concatBytes(Uint8Array.from([MULTIHASH_SHA2_256_CODE, MULTIHASH_SHA2_256_LEN]), digest);
    const cidBytes = concatBytes(encodeVarint(CID_V1), encodeVarint(multicodecCode), multihash);
    const cid = encodeMultibase58btc(cidBytes);
    return {
        cid,
        versionId: cid,
        canonicalJson,
        digestHex: toHex(digest),
    };
}
export function canonicalizeClaimsForCid(claims) {
    const stripped = { ...(claims || {}) };
    delete stripped['@context'];
    delete stripped['@type'];
    delete stripped['@id'];
    const normalized = canonicalizeValue(stripped, {
        stripMetaVersionId: false,
        stripNarrativeText: false,
        stripNestedElementIds: false,
    }, 0);
    return JSON.stringify(normalized);
}
export function claimsToCid(claims) {
    const canonicalJson = canonicalizeClaimsForCid(claims);
    const cidData = buildCidV1FromCanonicalJson(canonicalJson, DEFAULT_MULTICODEC_DAG_JSON);
    return {
        cid: cidData.cid,
        canonicalJson,
        digestHex: cidData.digestHex,
    };
}
export function assignCidToClaimsId(claims) {
    const out = JSON.parse(JSON.stringify(claims || {}));
    const { cid } = claimsToCid(out);
    out['@id'] = cid;
    return {
        claims: out,
        cid,
    };
}
export function fhirResourceToCid(resource, options = {}) {
    const canonicalJson = canonicalizeFhirResource(resource, options.canonicalization);
    return buildCidV1FromCanonicalJson(canonicalJson, options.multicodecCode ?? DEFAULT_MULTICODEC_DAG_JSON);
}
export function assignCidToFhirResourceVersionId(resource, options = {}) {
    const cid = fhirResourceToCid(resource, options);
    const out = JSON.parse(JSON.stringify(resource));
    const meta = (out.meta && typeof out.meta === 'object' && !Array.isArray(out.meta))
        ? { ...out.meta }
        : {};
    meta.versionId = cid.versionId;
    out.meta = meta;
    return {
        resource: out,
        mapping: {
            resourceType: String(out.resourceType || ''),
            resourceId: out.id ? String(out.id) : undefined,
            cid: cid.cid,
            versionId: cid.versionId,
        },
    };
}
export function assignCidToFhirBundleEntries(bundle, options = {}) {
    const out = JSON.parse(JSON.stringify(bundle));
    const entries = Array.isArray(out.entry) ? out.entry : [];
    const mappings = [];
    for (const entry of entries) {
        const resource = entry?.resource;
        if (!resource || typeof resource !== 'object' || Array.isArray(resource))
            continue;
        const assigned = assignCidToFhirResourceVersionId(resource, options);
        entry.resource = assigned.resource;
        assigned.mapping.fullUrl = entry.fullUrl ? String(entry.fullUrl) : undefined;
        mappings.push(assigned.mapping);
    }
    return { bundle: out, mappings };
}
