// gdc-common-utils-ts/dist/utils/vp-token.js
import { Content } from './content.js';
function fallbackId() {
    const rand = Math.random().toString(36).slice(2, 10);
    return `id-${Date.now()}-${rand}`;
}
export function generateUuidLike() {
    const fn = globalThis?.crypto?.randomUUID;
    if (typeof fn === 'function')
        return fn.call(globalThis.crypto);
    return fallbackId();
}
export function buildEpochWindow(ttlSeconds = 300) {
    const iat = Math.floor(Date.now() / 1000);
    return { iat, exp: iat + Math.max(1, Math.floor(ttlSeconds)) };
}
export function createVP(input) {
    const ttl = input?.exp && input?.iat ? undefined : buildEpochWindow(300);
    const jti = input?.jti || generateUuidLike();
    const nonce = input?.nonce || generateUuidLike();
    const base = {
        iss: String(input?.iss || ''),
        sub: input?.sub,
        aud: input?.aud,
        jti,
        iat: input?.iat ?? ttl?.iat,
        exp: input?.exp ?? ttl?.exp,
        nonce,
        vp: {
            '@context': ['https://www.w3.org/2018/credentials/v1'],
            type: ['VerifiablePresentation'],
            holder: input?.vp?.holder || input?.iss || '',
            verifiableCredential: [],
            ...(input?.vp || {}),
        },
    };
    return base;
}
export function addVC(vpPayload, vcJwt) {
    const v = String(vcJwt || '').trim();
    if (!v)
        return vpPayload;
    vpPayload.vp.verifiableCredential.push(v);
    return vpPayload;
}
export function prepareForSignature(header, payload) {
    const encodedHeader = Content.objectToRawBase64UrlSafe(header);
    const encodedPayload = Content.objectToRawBase64UrlSafe(payload);
    return {
        encodedHeader,
        encodedPayload,
        signingInput: `${encodedHeader}.${encodedPayload}`,
    };
}
export function prepareBytesForSignature(header, payload) {
    const { signingInput } = prepareForSignature(header, payload);
    return new TextEncoder().encode(signingInput);
}
export function buildVpTokenCompact(encodedHeader, encodedPayload, signatureBase64Url) {
    return `${encodedHeader}.${encodedPayload}.${String(signatureBase64Url || '').trim()}`;
}
