// gdc-common-utils-ts/dist/models/clinical-sections.js
/**
 * Canonical registry of supported clinical document sections (e.g., IPS Composition sections).
 *
 * Design goals:
 * - Use stable codes (LOINC) for interoperability.
 * - Provide a reverse-DNS i18n key for consistent UI translations across apps.
 * - Keep this in `gdc-common-utils-ts` so both backend and frontend can share it.
 */
import { clinicalSectionTitleEn } from './clinical-sections.en.js';
export const LOINC_SYSTEM_URL = 'http://loinc.org';
export const LOINC_SYSTEM_REVERSE_DNS = 'org.loinc';
export function loincI18nKey(code) {
    return `${LOINC_SYSTEM_REVERSE_DNS}.${code}`;
}
export const supportedClinicalSectionCodes = Object.freeze(Object.keys(clinicalSectionTitleEn));
export const clinicalSectionRegistry = (() => {
    const entries = Object.entries(clinicalSectionTitleEn).map(([code, titleEn]) => {
        const descriptor = {
            system: LOINC_SYSTEM_URL,
            code,
            i18nKey: loincI18nKey(code),
            titleEn,
            ips: true,
        };
        return [code, descriptor];
    });
    return Object.freeze(Object.fromEntries(entries));
})();
export { clinicalSectionTitleEn };
export function getClinicalSectionByCode(code) {
    return clinicalSectionRegistry[code];
}
