// gdc-common-utils-ts/dist/models/confidential-job.js
// Copyright 2025 Antifraud Services Inc. under the Apache License, Version 2.0.
// File: crypto-ts/models/confidential-job.ts
/**
 * Defines the possible statuses of a job throughout its lifecycle.
 */
export var JobStatus;
(function (JobStatus) {
    /** The job has been created locally but not yet submitted. */
    JobStatus["DRAFT"] = "DRAFT";
    /** The job is in the process of being submitted to the server. */
    JobStatus["SUBMITTING"] = "SUBMITTING";
    /** The job was successfully submitted and is awaiting asynchronous processing. */
    JobStatus["SENT"] = "SENT";
    /** The server has finished processing the job and returned a final result. This is a terminal state. */
    JobStatus["COMPLETED"] = "COMPLETED";
    /** The job failed due to a transport-level error and will not be retried. This is a terminal state. */
    JobStatus["FAILED"] = "FAILED";
    /** The job failed due to a transient error and may be retried. */
    JobStatus["ERROR_RETRYABLE"] = "ERROR_RETRYABLE";
})(JobStatus || (JobStatus = {}));
