// gdc-common-utils-ts/dist/models/consent-rule.js
// src/models/consent-rule.ts
export var ClaimConsent;
(function (ClaimConsent) {
    ClaimConsent["decision"] = "Consent.decision";
    ClaimConsent["action"] = "Consent.action";
    ClaimConsent["category"] = "Consent.category";
    ClaimConsent["subject"] = "Consent.subject";
    ClaimConsent["actorIdentifier"] = "Consent.actor-identifier";
    ClaimConsent["actorRole"] = "Consent.actor-role";
    ClaimConsent["date"] = "Consent.date";
    ClaimConsent["periodStart"] = "Consent.period-start";
    ClaimConsent["periodEnd"] = "Consent.period-end";
    ClaimConsent["grantee"] = "Consent.grantee";
    ClaimConsent["verifiedBy"] = "Consent.verified-by";
    ClaimConsent["verifiedDate"] = "Consent.verified-date";
    ClaimConsent["purpose"] = "Consent.purpose";
    ClaimConsent["identifier"] = "Consent.identifier";
    ClaimConsent["attachmentContentType"] = "Consent.attachment-contentType";
    ClaimConsent["attachmentData"] = "Consent.attachment-data";
    ClaimConsent["attachmentId"] = "Consent.attachment-id";
})(ClaimConsent || (ClaimConsent = {}));
