// gdc-common-utils-ts/dist/models/urlPath.js
// Copyright 2025 Antifraud Services Inc. under the Apache License, Version 2.0.
// File: crypto-ts/models/urlPath.ts
/**
 * Defines the standardized business sectors supported by the gateway.
 * Using an enum ensures type safety and prevents the use of arbitrary strings.
 */
export var Sector;
(function (Sector) {
    Sector["TEST"] = "test";
    Sector["SYSTEM"] = "system";
    Sector["HEALTH_CARE"] = "health-care";
    Sector["HEALTH_INSURANCE"] = "health-insurance";
    Sector["EMERGENCY"] = "emergency";
    Sector["RESEARCH"] = "research";
})(Sector || (Sector = {}));
export var Section;
(function (Section) {
    /** Managing registration of organizations */
    Section["registry"] = "registry";
    /** Managing data of the hosted organization */
    Section["entity"] = "entity";
    /** Managing data of the hosted individual */
    Section["individual"] = "individual";
    /** Managing data in the blockchain network */
    Section["network"] = "network";
})(Section || (Section = {}));
/** Standards, specifications and formats for data supported in the url path */
export var Format;
(function (Format) {
    Format["Schema"] = "org.schema";
    Format["FhirApi"] = "org.hl7.fhir.api";
    //Pdf' = 'pdf',
})(Format || (Format = {}));
/** Types of resources supported in the url path */
export var Resource;
(function (Resource) {
    Resource["Person"] = "Person";
    Resource["RelatedPerson"] = "RelatedPerson";
    Resource["Employee"] = "Emloyee";
    Resource["EmployeeRole"] = "EmloyeeRole";
    Resource["Practitioner"] = "Practitioner";
    Resource["PractitionerRole"] = "PractitionerRole";
    Resource["Organization"] = "Organization";
    Resource["Location"] = "Location";
    Resource["Group"] = "Group";
})(Resource || (Resource = {}));
export var JobAction;
(function (JobAction) {
    JobAction["_batch"] = "_batch";
    JobAction["_create"] = "_create";
    JobAction["_discovery"] = "_discovery";
})(JobAction || (JobAction = {}));
export var knownDomainsReversedEnum;
(function (knownDomainsReversedEnum) {
    knownDomainsReversedEnum["org.schema"] = "org.schema";
    knownDomainsReversedEnum["org.hl7.fhir.api"] = "org.hl7.fhir.api";
    knownDomainsReversedEnum["org.hl7.fhir.r4"] = "org.hl7.fhir.r4";
    knownDomainsReversedEnum["org.ilo.isco"] = "org.ilo.isco";
    knownDomainsReversedEnum["net.openid"] = "net.openid";
    // Add other known standards here
})(knownDomainsReversedEnum || (knownDomainsReversedEnum = {}));
;
/**
 * A list of known, fully-qualified context prefixes in reverse DNS format.
 * This is used by the claim normalization utility to identify claims that
 * are already interoperable and should not be modified.
 * All entries should be in lowercase.
 */
export const knownDomainsReversed = [
    knownDomainsReversedEnum["org.schema"],
    knownDomainsReversedEnum["org.hl7.fhir.api"],
    knownDomainsReversedEnum["org.hl7.fhir.r4"],
    knownDomainsReversedEnum["org.ilo.isco"],
    knownDomainsReversedEnum["net.openid"]
    // Add other known standards here
];
